src/sidebar/components/test/annotation-header-test.js

```diff
@@ -13,25 +13,41 @@ var fakeDocumentMeta = {
 
 describe('sidebar.components.annotation-header', function () {
   var $componentController;
+  var fakeFeatures;
   var fakeGroups;
+  var fakePersona;
   var fakeSettings = { usernameUrl: 'http://www.example.org/' };
   var fakeServiceUrl;
 
-  before(function () {
+  beforeEach('Initialize fakePersona', () => {
+    fakePersona = {
```

Comment:
I probably would have chosen not to stub out the simple utility functions in `persona.js` but rather have a couple of annotation fixtures where the third party (or not) status is determined from the username. The danger with stubbing so much is that you create additional ways in which the code could be wrong (eg. passing the wrong values to these functions) but still pass, or you have to add a whole bunch of extra tests to validate their usage - and we've already got >100 LOC of test code for 9 lines of actual logic. That's admirably thorough but it is also stuff that has to be maintained.

Reply:
The 100 lines of tests for `displayName()` is really just 20 lines of test code, and 80 lines of declarative test code objects, and I don't think the tests _not_ faking persona would reduce the necessity to test this number of cases? They are the possible routes through the internal logic of `displayName()`, given the possible values its params and dependencies can take. Replacing these 10 test cases with just a couple of fixtures would simply mean that not all the combinations were being tested, unless I'm misunderstanding?

I think this is an important difference between us, and between the h code and the client code, in test and code design philosophy, and I recall we've discussed it before.

I actually think it's really important to fake out dependencies like `persona`, even (seemingly) simple ones. And I don't think it creates additional ways in which the code could be wrong, but the tests could still be passing. I think it takes already-existing ways that could happen and makes them explicit rather than implicit. **Stubbing itself is not a bad thing**. If you have _too much_ stubbing in your tests then that _is_ a code smell, but it    describe('#htmlLink()', function () {
      it('returns the HTML link when available', function () {
        var ann = fixtures.defaultAnnotation();
        ann.links = { html: 'https://annotation.service/123' };
        var ctrl = $componentController('annotationHeader', {}, {
          annotation: ann,
        });
        assert.equal(ctrl.htmlLink(), ann.links.html);
      });

      it('returns an empty string when no HTML link is available', function () {
        var ann = fixtures.defaultAnnotation();
        ann.links = {};
        var ctrl = $componentController('annotationHeader', {}, {
          annotation: ann,
        });
        assert.equal(ctrl.htmlLink(), '');
      });
    });

    describe('#documentMeta()', function () {
      it('returns the domain, title link and text for the annotation', function () {
        var ann = fixtures.defaultAnnotation();
        var ctrl = $componentController('annotationHeader', {}, {
          annotation: ann,
        });
        assert.deepEqual(ctrl.documentMeta(), fakeDocumentMeta);
      });
    });

    describe('#displayName', () => {
      [
        {
          context: 'when the api_render_user_info feature flag is turned off in h',
          it: 'returns the username',
          user_info: undefined,
          client_display_names: false,
          isThirdPartyUser: false,
          expectedResult: 'TEST_USERNAME',
        },
        {
          context: 'when the api_render_user_info feature flag is turned off in h',
          it: 'returns the username even if the client_display_names feature flag is on',
          user_info: undefined,
          client_display_names: true,
          isThirdPartyUser: false,
          expectedResult: 'TEST_USERNAME',
        },
        {
          context: 'when the client_display_names feature flag    describe('#htmlLink()', function () {
      it('returns the HTML link when available', function () {
        var ann = fixtures.defaultAnnotation();
        ann.links = { html: 'https://annotation.service/123' };
        var ctrl = $componentController('annotationHeader', {}, {
          annotation: ann,
        });
        assert.equal(ctrl.htmlLink(), ann.links.html);
      });

      it('returns an empty string when no HTML link is available', function () {
        var ann = fixtures.defaultAnnotation();
        ann.links = {};
        var ctrl = $componentController('annotationHeader', {}, {
          annotation: ann,
        });
        assert.equal(ctrl.htmlLink(), '');
      });
    });

    describe('#documentMeta()', function () {
      it('returns the domain, title link and text for the annotation', function () {
        var ann = fixtures.defaultAnnotation();
        var ctrl = $componentController('annotationHeader', {}, {
          annotation: ann,
        });
        assert.deepEqual(ctrl.documentMeta(), fakeDocumentMeta);
      });
    });

    describe('#displayName', () => {
      [
        {
          context: 'when the api_render_user_info feature flag is turned off in h',
          it: 'returns the username',
          user_info: undefined,
          client_display_names: false,
          isThirdPartyUser: false,
          expectedResult: 'TEST_USERNAME',
        },
        {
          context: 'when the api_render_user_info feature flag is turned off in h',
          it: 'returns the username even if the client_display_names feature flag is on',
          user_info: undefined,
          client_display_names: true,
          isThirdPartyUser: false,
          expectedResult: 'TEST_USERNAME',
        },
        {
          context: 'when the client_display_names feature flag is off in h',
          it: 'returns the username',
          user_info: { display_name: null },
          client_display_names: false,
          isThirdPartyUser: false,
          expectedResult: 'TEST_USERNAME',
        },
        {
          context: 'when the client_display_names feature flag is off in h',
          it: 'returns the username even if the user has a display name',
          user_info: { display_name: 'Bill Jones' },
          client_display_names: false,
          isThirdPartyUser: false,
          expectedResult: 'TEST_USERNAME',
        },
        {
          context: 'when both feature flags api_render_user_info and ' +
                   'client_display_names are on',
          it: 'returns the username, if the user has no display_name',
          user_info: { display_name: null },
          client_display_names: true,
          isThirdPartyUser: false,
          expectedResult: 'TEST_USERNAME',
        },
        {
          context: 'when both feature flags api_render_user_info and ' +
                   'client_display_names are on',
          it: 'returns the display_name, if the user has one',
          user_info: { display_name: 'Bill Jones' },
          client_display_names: true,
          isThirdPartyUser: false,
          expectedResult: 'Bill Jones',
        },
        {
          context: 'when the client_display_names feature flag is off but ' +
                   'the user is a third-party user',
          it: 'returns display_name even though client_display_names is off',
          user_info: { display_name: 'Bill Jones' },
          client_display_names: false,
          isThirdPartyUser: true,
          expectedResult: 'Bill Jones',
        },
        {
          context: 'when client_display_names is on and the user is a ' +
                   'third-party user',
          it: 'returns the display_name',
          user_info: { display_name: 'Bill Jones' },
          client_display_names: true,
          isThirdPartyUser: true,
          expectedResult: 'Bill Jones',
        },
        {
          context: 'when the user is a third-party user but the ' +
                   'api_render_user_info feature flag is turned off in h',
          it: 'returns the username',
          user_info: undefined,
          client_display_names: true,
          isThirdPartyUser: true,
          expectedResult: 'TEST_USERNAME',
        },
        {
          context: "when the user is a third-party user but doesn't have a " +
                   'display_name',
          it: 'returns the username',
          user_info: { display_name: null},
          client_display_names: true,
          isThirdPartyUser: true,
          expectedResult: 'TEST_USERNAME',
        },
 is off in h',
          it: 'returns the username',
          user_info: { display_name: null },
          client_display_names: false,
          isThirdPartyUser: false,
          expectedResult: 'TEST_USERNAME',
        },
        {
          context: 'when the client_display_names feature flag is off in h',
          it: 'returns the username even if the user has a display name',
          user_info: { display_name: 'Bill Jones' },
          client_display_names: false,
          isThirdPartyUser: false,
          expectedResult: 'TEST_USERNAME',
        },
        {
          context: 'when both feature flags api_render_user_info and ' +
                   'client_display_names are on',
          it: 'returns the username, if the user has no display_name',
          user_info: { display_name: null },
          client_display_names: true,
          isThirdPartyUser: false,
          expectedResult: 'TEST_USERNAME',
        },
        {
          context: 'when both feature flags api_render_user_info and ' +
                   'client_display_names are on',
          it: 'returns the display_name, if the user has one',
          user_info: { display_name: 'Bill Jones' },
          client_display_names: true,
          isThirdPartyUser: false,
          expectedResult: 'Bill Jones',
        },
        {
          context: 'when the client_display_names feature flag is off but ' +
                   'the user is a third-party user',
          it: 'returns display_name even though client_display_names is off',
          user_info: { display_name: 'Bill Jones' },
          client_display_names: false,
          isThirdPartyUser: true,
          expectedResult: 'Bill Jones',
        },
        {
          context: 'when client_display_names is on and the user is a ' +
                   'third-party user',
          it: 'returns the display_name',
          user_info: { display_name: 'Bill Jones' },
          client_display_names: true,
          isThirdPartyUser: true,
          expectedResult: 'Bill Jones',
        },    describe('#htmlLink()', function () {
      it('returns the HTML link when available', function () {
        var ann = fixtures.defaultAnnotation();
        ann.links = { html: 'https://annotation.service/123' };
        var ctrl = $componentController('annotationHeader', {}, {
          annotation: ann,
        });
        assert.equal(ctrl.htmlLink(), ann.links.html);
      });

      it('returns an empty string when no HTML link is available', function () {
        var ann = fixtures.defaultAnnotation();
        ann.links = {};
        var ctrl = $componentController('annotationHeader', {}, {
          annotation: ann,
        });
        assert.equal(ctrl.htmlLink(), '');
      });
    });

    describe('#documentMeta()', function () {
      it('returns the domain, title link and text for the annotation', function () {
        var ann = fixtures.defaultAnnotation();
        var ctrl = $componentController('annotationHeader', {}, {
          annotation: ann,
        });
        assert.deepEqual(ctrl.documentMeta(), fakeDocumentMeta);
      });
    });

    describe('#displayName', () => {
      [
        {
          context: 'when the api_render_user_info feature flag is turned off in h',
          it: 'returns the username',
          user_info: undefined,
          client_display_names: false,
          isThirdPartyUser: false,
          expectedResult: 'TEST_USERNAME',
        },
        {
          context: 'when the api_render_user_info feature flag is turned off in h',
          it: 'returns the username even if the client_display_names feature flag is on',
          user_info: undefined,
          client_display_names: true,
          isThirdPartyUser: false,
          expectedResult: 'TEST_USERNAME',
        },
        {
          context: 'when the client_display_names feature flag is off in h',
          it: 'returns the username',
          user_info: { display_name: null },
          client_display_names: false,
          isThirdPartyUser: false,
          expectedResult: 'TEST_USERNAME',
        },
        {
          context: 'when the client_display_names feature flag is off in h',
          it: 'returns the username even if the user has a display name',
          user_info: { display_name: 'Bill Jones' },
          client_display_names: false,
          isThirdPartyUser: false,
          expectedResult: 'TEST_USERNAME',
        },
        {
          context: 'when both feature flags api_render_user_info and ' +
                   'client_display_names are on',
          it: 'returns the username, if the user has no display_name',
          user_info: { display_name: null },
          client_display_names: true,
          isThirdPartyUser: false,
          expectedResult: 'TEST_USERNAME',
        },
        {
          context: 'when both feature flags api_render_user_info and ' +
                   'client_display_names are on',
          it: 'returns the display_name, if the user has one',
          user_info: { display_name: 'Bill Jones' },
          client_display_names: true,
          isThirdPartyUser: false,
          expectedResult: 'Bill Jones',
        },
        {
          context: 'when the client_display_names feature flag is off but ' +
                   'the user is a third-party user',
          it: 'returns display_name even though client_display_names is off',
          user_info: { display_name: 'Bill Jones' },
          client_display_names: false,
          isThirdPartyUser: true,
          expectedResult: 'Bill Jones',
        },
        {
          context: 'when client_display_names is on and the user is a ' +
                   'third-party user',
          it: 'returns the display_name',
          user_info: { display_name: 'Bill Jones' },
          client_display_names: true,
          isThirdPartyUser: true,
          expectedResult: 'Bill Jones',
        },
        {
          context: 'when the user is a third-party user but the ' +
                   'api_render_user_info feature flag is turned off in h',
          it: 'returns the username',
          user_info: undefined,
          client_display_names: true,
          isThirdPartyUser: true,
          expectedResult: 'TEST_USERNAME',
        },
        {
          context: "when the user is a third-party user but doesn't have a " +
                   'display_name',
          it: 'returns the username',
          user_info: { display_name: null},
          client_display_names: true,
          isThirdPartyUser: true,
          expectedResult: 'TEST_USERNAME',
        },

        {
          context: 'when the user is a third-party user but the ' +
                   'api_render_user_info feature flag is turned off in h',
          it: 'returns the username',
          user_info: undefined,
          client_display_names: true,
          isThirdPartyUser: true,
          expectedResult: 'TEST_USERNAME',
        },
        {
          context: "when the user is a third-party user but doesn't have a " +
                   'display_name',
          it: 'returns the username',
          user_info: { display_name: null},
          client_display_names: true,
          isThirdPartyUser: true,
          expectedResult: 'TEST_USERNAME',
        },
's a smell of a design problem with your code (too many dependencies), it's _not a problem with the tests_.  Let me try to convince you:

I believe that `persona.isThirdPartyUser()` returns `true` if the `hypothes.is` part in a given username param like `acct:seanh@hypothes.is` differs from the given authDomain param. This is an internal implementation detail of `persona`.

So if the `annotation-header` tests use the real `persona`, and I want to write different tests for when `persona.isThirdPartyUser()` returns `true` and when it returns `false`, then I need to write some tests where username is `acct:seanh@hypothes.is` and authDomain is `hypothes.is`, and some tests where username is `acct:seanh@hypothes.is` and authDomain is `somethingelse.com`. This implicitly encodes implementation details of how `persona` works and of what the format of Hypothesis userids is into the tests for `annotation-header`, and the tests for every other module that uses `persona` and doesn't mock it.

If the definition of what constitutes a third-party user is then changed in `persona`, then `annotation-header` tests (and many more tests of other modules that use `persona`) would be failing and all need to be fixed. Even worse, `annotation-header` tests like `"it returns the username for a third-party user with no display name"` could be _succeeding_ even though the code was wrong, e.g. the code was crashing, because the test would no longer be testing what it said it was testing, for example the params that the test sets it up for `annotation-header` to pass to `isThirdPartyUser()` would no longer be ones that constitute a third-party account, it could now be unintentionally testing a third-party account.

Multiply that sort of thing across a large test suite for an entire codebase as it evolves over time, and you've got a hell of a problem, and may not be easy to see the extent of the problem you have.

That's why I much prefer this sort of thing:

```javascript
    fakePersona = {
      isThirdPartyUser: sinon.stub().returns(false),
      username: sinon.stub().returns('TEST_USERNAME'),
    };
```    describe('#htmlLink()', function () {
      it('returns the HTML link when available', function () {
        var ann = fixtures.defaultAnnotation();
        ann.links = { html: 'https://annotation.service/123' };
        var ctrl = $componentController('annotationHeader', {}, {
          annotation: ann,
        });
        assert.equal(ctrl.htmlLink(), ann.links.html);
      });

      it('returns an empty string when no HTML link is available', function () {
        var ann = fixtures.defaultAnnotation();
        ann.links = {};
        var ctrl = $componentController('annotationHeader', {}, {
          annotation: ann,
        });
        assert.equal(ctrl.htmlLink(), '');
      });
    });

    describe('#documentMeta()', function () {
      it('returns the domain, title link and text for the annotation', function () {
        var ann = fixtures.defaultAnnotation();
        var ctrl = $componentController('annotationHeader', {}, {
          annotation: ann,
        });
        assert.deepEqual(ctrl.documentMeta(), fakeDocumentMeta);
      });
    });

    describe('#displayName', () => {
      [
        {
          context: 'when the api_render_user_info feature flag is turned off in h',
          it: 'returns the username',
          user_info: undefined,
          client_display_names: false,
          isThirdPartyUser: false,
          expectedResult: 'TEST_USERNAME',
        },
        {
          context: 'when the api_render_user_info feature flag is turned off in h',
          it: 'returns the username even if the client_display_names feature flag is on',
          user_info: undefined,
          client_display_names: true,
          isThirdPartyUser: false,
          expectedResult: 'TEST_USERNAME',
        },
        {
          context: 'when the client_display_names feature flag is off in h',
          it: 'returns the username',
          user_info: { display_name: null },
          client_display_names: false,
          isThirdPartyUser: false,
          expectedResult: 'TEST_USERNAME',
        },
        {
          context: 'when the client_display_names feature flag is off in h',
          it: 'returns the username even if the user has a display name',
          user_info: { display_name: 'Bill Jones' },
          client_display_names: false,
          isThirdPartyUser: false,
          expectedResult: 'TEST_USERNAME',
        },
        {
          context: 'when both feature flags api_render_user_info and ' +
                   'client_display_names are on',
          it: 'returns the username, if the user has no display_name',
          user_info: { display_name: null },
          client_display_names: true,
          isThirdPartyUser: false,
          expectedResult: 'TEST_USERNAME',
        },
        {
          context: 'when both feature flags api_render_user_info and ' +
                   'client_display_names are on',
          it: 'returns the display_name, if the user has one',
          user_info: { display_name: 'Bill Jones' },
          client_display_names: true,
          isThirdPartyUser: false,
          expectedResult: 'Bill Jones',
        },
        {
          context: 'when the client_display_names feature flag is off but ' +
                   'the user is a third-party user',
          it: 'returns display_name even though client_display_names is off',
          user_info: { display_name: 'Bill Jones' },
          client_display_names: false,
          isThirdPartyUser: true,
          expectedResult: 'Bill Jones',
        },
        {
          context: 'when client_display_names is on and the user is a ' +
                   'third-party user',
          it: 'returns the display_name',
          user_info: { display_name: 'Bill Jones' },
          client_display_names: true,
          isThirdPartyUser: true,
          expectedResult: 'Bill Jones',
        },
        {
          context: 'when the user is a third-party user but the ' +
                   'api_render_user_info feature flag is turned off in h',
          it: 'returns the username',
          user_info: undefined,
          client_display_names: true,
          isThirdPartyUser: true,
          expectedResult: 'TEST_USERNAME',
        },
        {
          context: "when the user is a third-party user but doesn't have a " +
                   'display_name',
          it: 'returns the username',
          user_info: { display_name: null},
          client_display_names: true,
          isThirdPartyUser: true,
          expectedResult: 'TEST_USERNAME',
        },


Because there are no internal implementation details of `persona` in that, and because rather than being implicit, `fakePersona` is an explicit statement of exactly what the `annotation-header` tests think is the public API that `persona` provides to `annotation-header`. Each `fakePersona` in the tests for a module that uses `persona` is always called `fakePersona`, and is always declared and initialised in one place at the top of the file, stubs are very simple and declarative, etc. So in the unfortunate event that a change does have to be made to `persona`'s API it's easy to find and update all the fakes of it.

There are other downsides to integrated tests to consider as well, such as that they're slower, that they make bugs harder to locate because a bug in `persona` can cause a cascade of test failures in other modules, that they make the tests harder to understand (you need to know implementation details of `persona` to understand what the `annotation-header` tests are doing), that they're more brittle and more likely to result in a large number of tests failing (or silently succeeding when they _should_ now be failing) as a result of some code change thus making code harder to change, isolated tests encourage TDD and better design, etc.

But the tight coupling of the tests of module A to implementation details of modules B, C and D is the one that worries me most, because over time it'll lead to an unmaintainable codebase.